`ob deploy push` crashes on GHC warnings

When pushing an obelisk project to the Amazon EC2 instance, any warning that was tolerated before now breaks the build:
```
<source file>:<line>:1: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
FAILED Running nix-build on /home/anton/orga/naproche/ob-nave-deploy/src [server.system]
ob: <stdout>: commitAndReleaseBuffer: invalid argument (invalid character)
```
This seems to be related to #335

An entire warning would look like:
```
/home/anton/orga/naproche/Naproche-SAD/Isabelle/src/Isabelle/Term_XML/Decode.hs:21:1: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘indexname’:
        Patterns not matched:
            []
            (_:_:_:_)
   |
21 | indexname [a] = (a, 0)
   | ^^^^^^^^^^^^^^^^^^^^^^...
```
where the last three lines are color-coded in blue and violet. Maybe that is the problem?

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -17,6 +17,7 @@ This project's release branch is `master`. This log is written from the perspect
   * [#931](https://github.com/obsidiansystems/obelisk/pull/931): Fix bug in `ob deploy init` where `ssh-keygen` was not found in nix store.
   * [#934](https://github.com/obsidiansystems/obelisk/pull/934)[#936](https://github.com/obsidiansystems/obelisk/pull/936): obelisk now always uses absolute paths when generating `.ghci` files for REPL and IDE support. This is a **BREAKING** change for some old tools that could not handle absolute paths properly. However, due to the behavior of cross-volume relative paths on certain platforms, such as modern MacOS, we cannot guarantee proper operation of obelisk with relative paths.
   * [#948](https://github.com/obsidiansystems/obelisk/pull/948): obelisk now always invokes bash instead of the system-wide shell when it can. Some sub-programs, like ghcid, will still use the system-wide shell, which can cause subtle problems due to impurities.
+  * [#949](https://github.com/obsidiansystems/obelisk/pull/949): obelisk now configures its output and error streams to transliterate unsupported to a known-good replacement character.
 * obelisk-route
   * [#915](https://github.com/obsidiansystems/obelisk/pull/915): Add routeLinkAttr to Obelisk.Route.Frontend. This allows the creation of route links with additional, user-specified attributes.
   * [#918](https://github.com/obsidiansystems/obelisk/pull/918): Add GHC 8.10.7 support for `obelisk-route`

--- a/lib/command/src/Obelisk/Command.hs
+++ b/lib/command/src/Obelisk/Command.hs
@@ -21,7 +21,8 @@ import System.Directory
 import System.Environment
 import System.FilePath
 import qualified System.Info
-import System.IO (hIsTerminalDevice, stdout, stderr, hGetEncoding, hSetEncoding, mkTextEncoding, textEncodingName)
+import System.IO (hIsTerminalDevice, Handle, stdout, stderr, hGetEncoding, hSetEncoding, mkTextEncoding)
+import GHC.IO.Encoding.Types (textEncodingName)
 import System.Process (rawSystem)
 
 import Obelisk.App
@@ -372,8 +373,8 @@ main' argsCfg = do
   -- TransliterateCodingFailure so that, on encodings which do not
   -- support our fancy characters, we print a replacement character
   -- instead of exploding.
-  hSetTranslit stdout
-  hSetTranslit stderr
+  liftIO $ hSetTranslit stdout
+  liftIO $ hSetTranslit stderr
 
   putLog Debug $ T.pack $ unwords
     [ "Starting Obelisk <" <> obPath <> ">"

--- a/lib/command/src/Obelisk/Command.hs
+++ b/lib/command/src/Obelisk/Command.hs
@@ -9,7 +9,7 @@ module Obelisk.Command where
 import Control.Monad.IO.Class (MonadIO, liftIO)
 import Data.Bool (bool)
 import Data.Foldable (for_)
-import Data.List (isInfixOf, isPrefixOf)
+import Data.List (isInfixOf, isPrefixOf, notElem)
 import Data.List.NonEmpty (NonEmpty, nonEmpty)
 import qualified Data.List.NonEmpty as NonEmpty
 import qualified Data.Map as Map
@@ -21,7 +21,7 @@ import System.Directory
 import System.Environment
 import System.FilePath
 import qualified System.Info
-import System.IO (hIsTerminalDevice, stdout)
+import System.IO (hIsTerminalDevice, stdout, stderr, hGetEncoding, hSetEncoding, mkTextEncoding, textEncodingName)
 import System.Process (rawSystem)
 
 import Obelisk.App
@@ -351,11 +351,30 @@ runCommand f = flip runObelisk f =<< mkObeliskConfig
 main :: IO ()
 main = runCommand . main' =<< getArgsConfig
 
+-- | Change the character encoding of the given Handle to transliterate
+-- on unsupported characters, instead of throwing an exception.
+hSetTranslit :: Handle -> IO ()
+hSetTranslit h = do
+  menc <- hGetEncoding h
+  case fmap textEncodingName menc of
+    Just name | '/' `notElem` name -> do
+      enc' <- mkTextEncoding $ name ++ "//TRANSLIT"
+      hSetEncoding h enc'
+    _ -> return ()
+
 main' :: MonadObelisk m => ArgsConfig -> m ()
 main' argsCfg = do
   obPath <- liftIO getExecutablePath
   myArgs <- liftIO getArgs
   logLevel <- getLogLevel
+
+  -- NB: We set the standard output and standard error streams to
+  -- TransliterateCodingFailure so that, on encodings which do not
+  -- support our fancy characters, we print a replacement character
+  -- instead of exploding.
+  hSetTranslit stdout
+  hSetTranslit stderr
+
   putLog Debug $ T.pack $ unwords
     [ "Starting Obelisk <" <> obPath <> ">"
     , "args=" <> show myArgs

--- a/lib/command/src/Obelisk/Command.hs
+++ b/lib/command/src/Obelisk/Command.hs
@@ -353,7 +353,7 @@ main :: IO ()
 main = runCommand . main' =<< getArgsConfig
 
 -- | Change the character encoding of the given Handle to transliterate
--- on unsupported characters, instead of throwing an exception.
+-- unsupported characters, instead of throwing an exception.
 hSetTranslit :: Handle -> IO ()
 hSetTranslit h = do
   menc <- hGetEncoding h

